### PR TITLE
Touch control fixes

### PR DESCRIPTION
**Pull request by selaux**
This should fix:

- #1679: Caused by the mouse region deleted within its callback, leading to a segfault. We now guard against deleted / disabled regions in `MouseCallbackPrimarySecondary` and `ButtonCallbackPrimarySecondary` before calling the callbacks
- #1666: We now always compare callback reason flags using the `&` operator
- Assertion messages on tactical placement screen: The array was not properly zeroed out when resetting merc placements

### CODE REVIEW

**Review by momoko-h — COMMENTED** [review]
I dislike this change. I'd rather remove the assertion in MSYS_DefineRegion, because its making the wrong assumption that the memory block it's initializing was a `MOUSE_REGION` before, so all it is testing is whether the previous owner had set some random bit. It's basically a pre-C++ version of a constructor, and as such must not assume zero initialized memory.

In the original code the `AssertMsg` was inside an `#ifdef MOUSESYSTEM_DEBUGGING` anyway.

File: src/sgp/Button_System.h [hunk 1]
```diff
@@ -41,6 +40,13 @@ struct GUI_BUTTON;
 // GUI_BUTTON callback function type
 typedef std::function<void(GUI_BUTTON*, UINT32)> GUI_CALLBACK;
 
+std::function<void(GUI_BUTTON*, UINT32)> ButtonCallbackPrimarySecondary(
+	std::function<void(GUI_BUTTON*, UINT32)> primaryAction,
+	std::function<void(GUI_BUTTON*, UINT32)> secondaryAction,
+	std::function<void(GUI_BUTTON*, UINT32)> allEvents = nullptr,
+	bool triggerPrimaryOnMouseDown = false
+);
+
```
Comment (momoko-h) [inline src/sgp/Button_System.h:49]: primaryAction, secondaryAction and allEvents should probably use GUI_CALLBACK.

File: src/sgp/MouseSystem.cc [hunk 2]
```diff
@@ -974,3 +974,13 @@ void MOUSE_REGION::AllowDisabledRegionFastHelp(bool const allow)
...
 }
+
+std::function<void(MOUSE_REGION*, UINT32)> MouseCallbackPrimarySecondary(
+	std::function<void(MOUSE_REGION*, UINT32)> primaryAction,
+	std::function<void(MOUSE_REGION*, UINT32)> secondaryAction,
+	std::function<void(MOUSE_REGION*, UINT32)> allEvents,
+	bool triggerPrimaryOnMouseDown
+)
+{
+	return CallbackPrimarySecondary<MOUSE_REGION, MSYS_REGION_ENABLED>(primaryAction, secondaryAction, allEvents, triggerPrimaryOnMouseDown);
```
Comment (momoko-h) [inline src/sgp/MouseSystem.cc:985]: MOUSE_CALLBACK for the parameter types?

**Comment by momoko-h** [comment]
The `gfItemPopupRegionCallbackEndFix` based workarounds in Interface_Items.cc seem to be no longer needed?

**Comment by selaux** [comment]
Good catch regarding `gfItemPopupRegionCallbackEndFix`. I removed the related code.

Regarding `MSYS_DefineRegion` I don't have a strong oppinion. From my understanding it was less a check for the specific flag, but for a zeroed out memory region. But I agree that this only tests for basically a single bit, so it can probably be removed.

**Comment by lynxlynxlynx** [comment]
The less code the better and since this was ifdefed out before, I see no reason to keep it.

**Pull request merged**